jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/StageCompletionConditionTest.java

```diff
@@ -0,0 +1,362 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jbpm.casemgmt.impl;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.drools.compiler.kie.builder.impl.InternalKieModule;
+import org.jbpm.casemgmt.api.dynamic.TaskSpecification;
+import org.jbpm.casemgmt.api.model.CaseStatus;
+import org.jbpm.casemgmt.api.model.instance.CaseFileInstance;
+import org.jbpm.casemgmt.api.model.instance.CaseInstance;
+import org.jbpm.casemgmt.impl.model.instance.CaseInstanceImpl;
+import org.jbpm.casemgmt.impl.util.AbstractCaseServicesBaseTest;
+import org.jbpm.kie.services.impl.KModuleDeploymentUnit;
+import org.jbpm.runtime.manager.impl.deploy.DeploymentDescriptorImpl;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.kie.api.KieServices;
+import org.kie.api.builder.ReleaseId;
+import org.kie.api.task.model.TaskSummary;
+import org.kie.internal.query.QueryContext;
+import org.kie.internal.query.QueryFilter;
+import org.kie.internal.runtime.conf.DeploymentDescriptor;
+import org.kie.internal.runtime.conf.NamedObjectModel;
+import org.kie.internal.runtime.conf.RuntimeStrategy;
+import org.kie.scanner.MavenRepository;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.kie.scanner.MavenRepository.*;
+
+public class StageCompletionConditionTest extends AbstractCaseServicesBaseTest {
+
+    private static final String STAGE_WITH_TASK_AUTOCOMPLETE = "StageWithTaskAutocomplete";
+    private static final String STAGE_WITH_TASK_COUNT = "StageWithTaskCount";
+    private static final String STAGE_WITH_TASK_NO_AUTOSTART = "StageWithTaskNoAutoStart";
+    private static final String STAGE_WITH_TASK_PROCESS_VARIABLE = "StageWithTaskProcessVariable";
+    private static final String STAGE_WITH_TASK_CASE_FILE_VARIABLE = "StageWithTaskCaseFileVariable";
+    private static final String STAGE_WITH_TASK_CASE_FILE_VARIABLE_NO_PREFIX = "StageWithTaskCaseFileVariableNoPrefix";
+    private static final String STAGE_WITH_TASK_CASE_DATA_DEFINITION_ID = "StageWithTaskCaseDataDefinitionId";
+    private static final String STAGE_WITH_TASK_CASE_DATA_VARIABLE = "StageWithTaskCaseDataVariable";
+    private static final String STAGE_WITH_TASK_CASE_FILE_AND_PROCESS_VARIABLE = "StageWithTaskCaseFileAndProcessVariable";
+    private static final String USER_TASK_PROCESS = "UserTask";
+
+    private static final String INSIDE_TASK = "InsideTask";
+    private static final String DYNAMIC_TASK = "DynamicTask";
+    private static final String HELLO_TASK = "Hello";
+
+    private static final String USER = "john";
+
+    private KModuleDeploymentUnit deploymentUnit;
+
+    @Before
+    public void prepare() throws Exception {
```

Comment:
The method prepares several different things that can be refactored into standalone methods in a parent class and reused in other test cases (if those methods are properly parameterized). Now is has two level of abstraction, as it orchestrates all steps of preparation and deals with details of every step.